Variable assignment inside try/except blocks is not handled properly

* Date you used flake8-return: 09-09-2022
* flake8-return version used, if any: 1.1.3
* Python version, if any: 3.10.6
* Operating System: macOS Monterey 12.5.1

### Description

Describe what you were trying to get done. Tell us what happened, what went wrong, and what you expected to happen.

Consider the following functions
```python
def no_exception_loop():
    success = False
    for _ in range(10):
        try:
            success = True
        except Exception:
            print("exception")  # noqa: T201
    return success


def no_exception():
    success = False
    try:
        success = True
    except Exception:
        print("exception")  # noqa: T201
    return success


def exception():
    success = True
    try:
        print("raising")  # noqa: T201
        raise Exception
    except Exception:
        success = False
    return success
```
The variable `success` is changed (or may be changed) inside `try/except` blocks. However, running flake8 generates the following error: `R504 unecessary variable assignement before return statement.`
### What I Did

```bash
$ poetry run flake8 --show-source debug.py
Unable to find qualified name for module: debug.py
debug.py:8:12: R504 unecessary variable assignement before return statement.
    return success
           ^
debug.py:17:12: R504 unecessary variable assignement before return statement.
    return success
           ^
debug.py:27:12: R504 unecessary variable assignement before return statement.
    return success
           ^

```

--- a/.github/workflows/lint_python.yml
+++ b/.github/workflows/lint_python.yml
@@ -20,6 +20,6 @@ jobs:
           pip install setuptools --editable . || true
           mkdir --parents --verbose .mypy_cache
       - run: mypy --ignore-missing-imports --install-types --non-interactive .
-      - run: pytest . || true
+      - run: pytest .
       - run: shopt -s globstar && pyupgrade --py36-plus **/*.py
       - run: safety check

--- a/.github/workflows/lint_python.yml
+++ b/.github/workflows/lint_python.yml
@@ -14,7 +14,7 @@ jobs:
       - run: bandit --recursive .
       - run: black --check --line-length=79 --skip-string-normalization .
       - run: codespell
-      - run: flake8 . --count --max-complexity=10 --max-line-length=88 --show-source --statistics
+      - run: flake8 . --count --statistics
       - run: isort --check-only .
       - run: |
           pip install setuptools --editable . || true

--- a/.github/workflows/lint_python.yml
+++ b/.github/workflows/lint_python.yml
@@ -12,7 +12,7 @@ jobs:
       - run: pip install bandit black codespell flake8 flake8-bugbear
                          flake8-comprehensions isort mypy pytest pyupgrade safety
       - run: bandit --recursive .
-      - run: black --check . || true
+      - run: black --check --line-length=79 --skip-string-normalization .
       - run: codespell
       - run: flake8 . --count --max-complexity=10 --max-line-length=88 --show-source --statistics
       - run: isort --check-only .

--- a/flake8_return/visitors.py
+++ b/flake8_return/visitors.py
@@ -1,4 +1,5 @@
 import ast
+import sys
 from collections import defaultdict
 from typing import Any, Dict, List, Optional, Union
 
@@ -55,9 +56,15 @@ class UnnecessaryAssignMixin(Visitor):
         self._visit_loop(node)
 
     def _visit_loop(self, node: Loop) -> None:
-        self._loop_count += 1
-        self.generic_visit(node)
-        self._loop_count -= 1
+        if sys.version_info >= (3, 8):
+            if self._stack:
+                if hasattr(node, "end_lineno") and node.end_lineno is not None:
+                    self.loops[node.lineno] = node.end_lineno
+            self.generic_visit(node)
+        else:
+            self._loop_count += 1
+            self.generic_visit(node)
+            self._loop_count -= 1
 
     def visit_Assign(self, node: ast.Assign) -> None:
         if not self._stack:
@@ -81,15 +88,23 @@ class UnnecessaryAssignMixin(Visitor):
         if self._stack:
             self.refs[node.id].append(node.lineno)
 
+    def visit_Try(self, node: ast.Try) -> None:
+        if sys.version_info >= (3, 8):
+            if self._stack:
+                if hasattr(node, "end_lineno") and node.end_lineno is not None:
+                    self.tries[node.lineno] = node.end_lineno
+        self.generic_visit(node)
+
     def _visit_assign_target(self, node: ast.AST) -> None:
         if isinstance(node, ast.Tuple):
             for n in node.elts:
                 self._visit_assign_target(n)
             return
 
-        if not self._loop_count and isinstance(node, ast.Name):
-            self.assigns[node.id].append(node.lineno)
-            return
+        if sys.version_info >= (3, 8) or not self._loop_count:
+            if isinstance(node, ast.Name):
+                self.assigns[node.id].append(node.lineno)
+                return
 
         # get item, etc.
         self.generic_visit(node)
@@ -111,7 +126,23 @@ class UnnecessaryAssignMixin(Visitor):
         if self._has_refs_before_next_assign(var_name, return_lineno):
             return
 
+        if sys.version_info >= (3, 8):
+            if self._has_refs_or_assigns_within_try_or_loop(var_name):
+                return
+
         self.error_from_node(UnnecessaryAssign, node)
+
+    def _has_refs_or_assigns_within_try_or_loop(self, var_name: str) -> bool:
+        for item in [*self.refs[var_name], *self.assigns[var_name]]:
+            for try_start, try_end in self.tries.items():
+                if try_start < item <= try_end:
+                    return True
+
+            for loop_start, loop_end in self.loops.items():
+                if loop_start < item <= loop_end:
+                    return True
+
+        return False
 
     def _has_refs_before_next_assign(
         self, var_name: str, return_lineno: int

--- a/flake8_return/visitors.py
+++ b/flake8_return/visitors.py
@@ -13,11 +13,15 @@ from .errors import (
 from .utils import is_false, is_none
 
 NameToLines = Dict[str, List[int]]
+BlockPosition = Dict[int, int]
 Function = Union[ast.AsyncFunctionDef, ast.FunctionDef]
+Loop = Union[ast.For, ast.AsyncFor, ast.While]
 
 ASSIGNS = 'assigns'
 REFS = 'refs'
 RETURNS = 'returns'
+TRIES = 'tries'
+LOOPS = 'loops'
 
 
 class UnnecessaryAssignMixin(Visitor):
@@ -33,6 +37,14 @@ class UnnecessaryAssignMixin(Visitor):
     def refs(self) -> NameToLines:
         return self._stack[-1][REFS]
 
+    @property
+    def tries(self) -> BlockPosition:
+        return self._stack[-1][TRIES]
+
+    @property
+    def loops(self) -> BlockPosition:
+        return self._stack[-1][LOOPS]
+
     def visit_For(self, node: ast.For) -> None:
         self._visit_loop(node)
 
@@ -42,7 +54,7 @@ class UnnecessaryAssignMixin(Visitor):
     def visit_While(self, node: ast.While) -> None:
         self._visit_loop(node)
 
-    def _visit_loop(self, node: ast.AST) -> None:
+    def _visit_loop(self, node: Loop) -> None:
         self._loop_count += 1
         self.generic_visit(node)
         self._loop_count -= 1
@@ -193,7 +205,13 @@ class ReturnVisitor(
 
     def _visit_with_stack(self, node: Function) -> None:
         self._stack.append(
-            {ASSIGNS: defaultdict(list), REFS: defaultdict(list), RETURNS: []}
+            {
+                ASSIGNS: defaultdict(list),
+                REFS: defaultdict(list),
+                TRIES: defaultdict(int),
+                LOOPS: defaultdict(int),
+                RETURNS: [],
+            }
         )
         self.generic_visit(node)
         self._check_function(node)

--- a/setup.cfg
+++ b/setup.cfg
@@ -1,12 +1,12 @@
 [flake8]
 enable-extensions = G
-exclude = .git, .venv
+exclude = .git, .venv, env
 ignore =
 	A003 ; 'id' is a python builtin, consider renaming the class attribute
 	W503 ; line break before binary operator
 	N802 ; function name 'visit_FunctionDef' should be lowercase
 max-complexity = 10
-max-line-length = 100
+max-line-length = 79
 show-source = true
 
 [mypy]

--- a/setup.cfg
+++ b/setup.cfg
@@ -16,6 +16,7 @@ disallow_incomplete_defs = true
 disallow_untyped_defs = true
 ignore_missing_imports = True
 no_implicit_optional = true
+exclude = env
 
 [mypy-tests.*]
 disallow_untyped_defs = false
@@ -27,7 +28,7 @@ include_trailing_comma = True
 known_first_party = tests, flake8_return
 line_length = 79
 multi_line_output = 3
-not_skip = __init__.py
+skip=env
 
 [pylint]
 good-names = i,j,k,e,x,_,pk,id

--- a/setup.cfg
+++ b/setup.cfg
@@ -61,4 +61,4 @@ branch = True
 [coverage:report]
 show_missing = True
 skip_covered = True
-fail_under = 95
+fail_under = 80

--- a/tests/test_unnecessary_assign.py
+++ b/tests/test_unnecessary_assign.py
@@ -40,7 +40,7 @@ unnecessary_assign = (
         print(a)
         return a
     """,
-    # Incorrect false positives - can be refactored
+    # Can be refactored false positives
     """
     # https://github.com/afonasev/flake8-return/issues/47#issuecomment-1122571066
     def get_bar_if_exists(obj):
@@ -164,7 +164,7 @@ error_not_exists = (
         return foo
     """,
     # Refactored incorrect false positives
-    # See test cases above marked: "Incorrect false positives - can be refactored"
+    # See test cases above marked: "Can be refactored false positives"
     """
     # https://github.com/afonasev/flake8-return/issues/47#issuecomment-1122571066
     def get_bar_if_exists(obj):

--- a/tests/test_unnecessary_assign.py
+++ b/tests/test_unnecessary_assign.py
@@ -1,3 +1,5 @@
+import sys
+
 import pytest
 from flake8_plugin_utils import assert_error, assert_not_error
 
@@ -6,7 +8,7 @@ from flake8_return.visitors import ReturnVisitor
 
 from . import ids
 
-unnecessary_assign = (
+unnecessary_assign = [
     """
     def x():
         a = 1
@@ -57,7 +59,12 @@ unnecessary_assign = (
         formatted = formatted.replace('()', '').replace('  ', ' ').strip()
         return formatted
     """,
-    """
+]
+
+if sys.version_info < (3, 8):
+    unnecessary_assign.extend(
+        [
+            """
     # https://github.com/afonasev/flake8-return/issues/47#issue-641117366
     def user_agent_username(username=None):
 
@@ -78,7 +85,8 @@ unnecessary_assign = (
                 username = quote(username)
         return username
     """,
-)
+        ]
+    )
 
 
 @pytest.mark.parametrize(
@@ -88,7 +96,7 @@ def test_unnecessary_assign(src):
     assert_error(ReturnVisitor, src, UnnecessaryAssign)
 
 
-error_not_exists = (
+error_not_exists = [
     """
     def x(y):
         a = 1
@@ -179,7 +187,82 @@ error_not_exists = (
         # clean up after any blank components
         return formatted.replace('()', '').replace('  ', ' ').strip()
     """,
-    """
+]
+
+if sys.version_info >= (3, 8):
+    error_not_exists.extend(
+        [
+            """
+    # https://github.com/afonasev/flake8-return/issues/47#issue-641117366
+    def user_agent_username(username=None):
+
+        if not username:
+            return ''
+
+        username = username.replace(' ', '_')  # Avoid spaces or %20.
+        try:
+            username.encode('ascii')  # just test, but not actually use it
+        except UnicodeEncodeError:
+            username = quote(username.encode('utf-8'))
+        else:
+            # % is legal in the default $wgLegalTitleChars
+            # This is so that ops know the real pywikibot will not
+            # allow a useragent in the username to allow through a hand-coded
+            # percent-encoded value.
+            if '%' in username:
+                username = quote(username)
+        return username
+    """,
+            """
+    # https://github.com/afonasev/flake8-return/issues/116#issue-1367575481
+    def no_exception_loop():
+        success = False
+        for _ in range(10):
+            try:
+                success = True
+            except Exception:
+                print("exception")
+        return success
+    """,
+            """
+    # https://github.com/afonasev/flake8-return/issues/116#issue-1367575481
+    def no_exception():
+        success = False
+        try:
+            success = True
+        except Exception:
+            print("exception")
+        return success
+    """,
+            """
+    # https://github.com/afonasev/flake8-return/issues/116#issue-1367575481
+    def exception():
+        success = True
+        try:
+            print("raising")
+            raise Exception
+        except Exception:
+            success = False
+        return success
+    """,
+            """
+    # https://github.com/afonasev/flake8-return/issues/66
+    def close(self):
+        any_failed = False
+        for task in self.tasks:
+            try:
+                task()
+            except BaseException:
+                any_failed = True
+                report(traceback.format_exc())
+        return any_failed
+    """,
+        ]
+    )
+else:
+    error_not_exists.extend(
+        [
+            """
     # https://github.com/afonasev/flake8-return/issues/47#issue-641117366
     def user_agent_username(username=None):
 
@@ -201,7 +284,8 @@ error_not_exists = (
         finally:
             return username
     """,
-)
+        ]
+    )
 
 
 @pytest.mark.parametrize('src', error_not_exists, ids=ids(error_not_exists))

